### PR TITLE
code-generator: rewrite hack/update-codecs.sh into reusable generate-{internal,}-groups.sh

### PR DESCRIPTION
**Pull request by sttts**
Generating everything for groups inside of an apiserver (with internal types) becomes:

```shell
generate-internal-groups.sh all "$(dirname ${BASH_SOURCE})/../../.." k8s.io/sample-apiserver/pkg/client k8s.io/sample-apiserver/pkg/apis k8s.io/sample-apiserver/pkg/apis wardle:v1alpha1
```

Generating everything for a CRD (versioned types) becomes:

```shell
generate-groups.sh all "$(dirname ${BASH_SOURCE})/../../.." k8s.io/sample-apiserver/pkg/client k8s.io/sample-apiserver/pkg/apis wardle:v1alpha1
```

This should cover the 90% percent use-case. For the other 10% this can be forked and adapted as needed.

Furthermore, we can put this into a Docker container. Then code-generator consumers can then do:

```shell
$ docker run -v $GOPATH:/go k8s.io/code-generator:1.8 generate-group.sh github.com/foo/bar example:v1
```

This is possibly only the first step towards a `code-generator` binary. For the later deeper generator changes are necessary (e.g. https://github.com/kubernetes/kubernetes/pull/53202) and hence the later is only feasible in 1.9. **This PR here in contrast, we can cherry-pick to 1.8.**

```release-note
Add generate-groups.sh and generate-internal-groups.sh to k8s.io/code-generator to easily run generators against CRD or User API Server types.
```

Fixes https://github.com/kubernetes/kubernetes/issues/48714.

### CODE REVIEW

**Comment by mrIncompetent** [comment]
This is wonderful! Very much needed!

**Comment by deads2k** [comment]
@eparis can you help review the bash?

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  local GROUPS_WITH_VERSIONS="$4" # groupA:v1,v2,groupB,v1,groupC:v2
+  shift 4
+
+  # enumerate group versions
+  local FQ_APIS=() # e.g. k8s.io/api/apps/v1
+  for GVs in ${GROUPS_WITH_VERSIONS}; do
+    IFS=: read G Vs <<<"${GVs}"
+
+    # enumerate versions
+    for V in ${Vs//,/ } do
```
Comment (munnerz) [inline staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: I think there's a missing `;` after `for V in ${Vs//,/ }` here. I get the following when running on OS X:

Adding it resolves the following:

`/Users/james/go/src/k8s.io/kubernetes/vendor/k8s.io/code-generator/hack/lib/codegen.sh:98: parse error near 'done'`

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+CODEGEN_PKG=${CODEGEN_PKG:-$(ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo "k8s.io/code-generator")}
+
+for gen in defaulter-gen conversion-gen client-gen lister-gen informer-gen deepcopy-gen conversion-gen; do
+  echo "Building ${gen}"
+  go install ${CODEGEN_PKG}/cmd/${gen}
```
Comment (hongchaodeng) [inline staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: This incurs side effect for source-ing this lib.

**Comment by munnerz** [comment]
It'd be great if this library could also expose the `--verify` flag of it's underlying generators. I've found this really useful in CI, to ensure that generated files are up to date in the repo.

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    --versioned-clientset-package ${OUTPUT_PKG}/clientset/versioned \
+    --internal-clientset-package ${OUTPUT_PKG}/clientset/internalversion \
+    --listers-package ${OUTPUT_PKG}/listers \
+    --output-package ${OUTPUT_PKG}/informers \
+    --output-base ${BASE} \
+    "$@"
+}
+
+function generate-groups() {
+  local BASE="$1" # the base packages are resolved against (empty or the vendor/ directory ${OUTPUT_PKG} is in)
```
Comment (hongchaodeng) [inline staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: I have to confess these vars are difficult to understand. Can you give some specific examples?

**Review by hongchaodeng — COMMENTED** [review]
@sttts 
Awesome work!
A shared docker container like `k8s.io/code-generator:1.8` would be very useful too!

**Comment by hongchaodeng** [comment]
@sttts
Can you assign to me? I can help test the script out.

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    # enumerate versions
+    for V in ${Vs//,/ }; do
+      ALL_FQ_APIS+=(${EXT_APIS_PKG}/${G}/${V})
+      EXT_FQ_APIS+=(${EXT_APIS_PKG}/${G}/${V})
+    done
+  done
+
+  echo "Generating deepcopy funcs"
+  ${GOPATH}/bin/deepcopy-gen -i $(join , "${ALL_FQ_APIS[@]}") -O zz_generated.deepcopy --bounding-dirs ${INT_APIS_PKG},${EXT_APIS_PKG} "$@"
```
Comment (hongchaodeng) [inline staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: Where is boilerplate flag defined?

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: indeed, missing ;

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: true. We should do this only on-demand and only once.

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: Added a demonstration to sample-apiserver

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: simplified this. BASE is usually not needed, only in the kube case where we have staging/ directories.

**Comment by sttts** [comment]
@munnerz @hongchaodeng addressed all comments. ptal.

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh [hunk 5]
```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
```
Comment (hongchaodeng) [inline staging/src/k8s.io/code-generator/hack/lib/codegen.sh]:
```
#!/usr/bin/env bash
```

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+CODEGEN_PKG=./$(dirname "${BASH_SOURCE}")/../..
```
Comment (hongchaodeng) [inline staging/src/k8s.io/code-generator/hack/lib/codegen.sh]:
```
(dirname "${BASH_SOURCE}")
```

This won't work on other shell like zsh. For me, the terminal exited.

Comment (hongchaodeng) [reply staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: Use `$0` instead?

**Comment by hongchaodeng** [comment]
I confirm that after fixing my above minor comments it works for me 👍  :
```
codegen::generate-groups all \
  github.com/coreos/etcd-operator/pkg/client \
  github.com/coreos/etcd-operator/pkg/apis/ etcd:v1beta2
```

**Comment by munnerz** [comment]
I can also confirm it's working nicely in https://github.com/kubernetes/kubernetes/pull/52753 😄

**Comment by hongchaodeng** [comment]
@sttts
The tests are failing.

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh (hunk 6 above)
Comment (hongchaodeng) [reply staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: Actually this var is only used in `codegen::ensure-generators-built`.
Why make it global exported VAR? Maybe make it local there?

File: staging/src/k8s.io/code-generator/hack/lib/codegen.sh [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+# build generators on-demand
+CODEGEN_BUILT=false
```
Comment (hongchaodeng) [inline staging/src/k8s.io/code-generator/hack/lib/codegen.sh]: Also is it possible not to keep this state in this library?
Couldn't user just call install explicitly outside.

**Comment by sttts** [comment]
@ericchiang @munnerz addressed your comments. Moreover, I have switch from the codegen.sh library approach to two executable scripts in `k8s.io/code-generator`.  This is more in-line with the future `code-generator` binary, if we decide to rewrite the shell scripts to Go.

From my side this is complete now to go into master and the be cherry-picked into 1.8. A possible Golang rewrite has to wait for 1.9.

**Comment by hongchaodeng** [comment]
lgtm if tests pass

**Comment by sttts** [comment]
@pwittrock after seeing your community meeting slides with `Support for bootstrapping CRD go clients, go controllers, ref docs?`, would be great to get your input on these kind of changes.

**Pull request merged**